Handling abrupt disconnection in Oscilloscope Activity

**Actual Behaviour**

The app crashes when PSLab is Abruptly removed.

**Expected Behaviour**

The app shouldn't crash. It should land back to Home.

**Steps to reproduce it**

Menu Drawer -> Applications -> Oscilloscope -> Disconnect the device

**LogCat for the issue**

```
java.lang.RuntimeException: Unable to start activity ComponentInfo{org.fossasia.pslab/org.fossasia.pslab.activity.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.hardware.usb.UsbDevice.getDeviceName()' on a null object reference
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2423)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2483)
	at android.app.ActivityThread.access$900(ActivityThread.java:153)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1349)
	at android.os.Handler.dispatchMessage(Handler.java:102)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5438)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:738)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:628)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.hardware.usb.UsbDevice.getDeviceName()' on a null object reference
	at android.os.Parcel.readException(Parcel.java:1626)
	at android.os.Parcel.readException(Parcel.java:1573)
	at android.hardware.usb.IUsbManager$Stub$Proxy.hasDevicePermission(IUsbManager.java:555)
	at android.hardware.usb.UsbManager.hasPermission(UsbManager.java:386)
	at org.fossasia.pslab.activity.MainActivity.onCreate(MainActivity.java:102)
	at android.app.Activity.performCreate(Activity.java:6303)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2376)
	... 9 more

```
**Screenshots of the issue**

> None

**Would you like to work on the issue?**

Yes

--- a/app/src/main/java/org/fossasia/pslab/activity/OscilloscopeActivity.java
+++ b/app/src/main/java/org/fossasia/pslab/activity/OscilloscopeActivity.java
@@ -1,14 +1,19 @@
 package org.fossasia.pslab.activity;
 
+
 import android.app.AlertDialog;
+import android.content.Context;
 import android.content.DialogInterface;
 import android.content.Intent;
+import android.content.SharedPreferences;
 import android.graphics.Color;
 import android.graphics.Point;
 import android.net.Uri;
 import android.os.AsyncTask;
 import android.os.Bundle;
 import android.os.Handler;
+import android.os.HandlerThread;
+import android.os.Looper;
 import android.support.annotation.IdRes;
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
@@ -43,6 +48,7 @@ import org.fossasia.pslab.R;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 
 /**
@@ -331,7 +337,6 @@ public class OscilloscopeActivity extends AppCompatActivity implements
         if (task != null) {
             task.cancel(true);
         }
-
         super.onDestroy();
     }
 
@@ -417,15 +422,20 @@ public class OscilloscopeActivity extends AppCompatActivity implements
 
         @Override
         protected Void doInBackground(String... params) {
-            analogInput = params[0];
-            //no. of samples and timegap still need to be determined
-            scienceLab.captureTraces(1, 1000, 10, analogInput, false, null);
-            HashMap<String,double[]> data = scienceLab.fetchTrace(1); //fetching data
-            double[] xData = data.get("x");
-            double[] yData = data.get("y");
-            entries = new ArrayList<Entry>();
-            for (int i = 0; i < xData.length; i++) {
-                entries.add(new Entry((float) xData[i], (float) yData[i]));
+            try {
+                analogInput = params[0];
+                //no. of samples and timegap still need to be determined
+                scienceLab.captureTraces(1, 1000, 10, analogInput, false, null);
+                HashMap<String, double[]> data = scienceLab.fetchTrace(1); //fetching data
+                double[] xData = data.get("x");
+                double[] yData = data.get("y");
+                entries = new ArrayList<Entry>();
+                for (int i = 0; i < xData.length; i++) {
+                    entries.add(new Entry((float) xData[i], (float) yData[i]));
+                }
+            }
+            catch (NullPointerException e){
+                cancel(true);
             }
             return null;
         }

--- a/app/src/main/java/org/fossasia/pslab/receivers/USBDetachReceiver.java
+++ b/app/src/main/java/org/fossasia/pslab/receivers/USBDetachReceiver.java
@@ -30,25 +30,29 @@ public class USBDetachReceiver extends BroadcastReceiver {
 
     @Override
     public void onReceive(Context context, Intent intent) {
-        if (UsbManager.ACTION_USB_DEVICE_DETACHED.equals(intent.getAction())) {
-            UsbDevice device = (UsbDevice) intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
-            if (device != null) {
-                ScienceLabCommon.scienceLab.close();
-                Toast.makeText(context, "USB Device Disconnected", Toast.LENGTH_SHORT).show();
+        try {
+            if (UsbManager.ACTION_USB_DEVICE_DETACHED.equals(intent.getAction())) {
+                UsbDevice device = (UsbDevice) intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
+                if (device != null) {
+                    ScienceLabCommon.scienceLab.close();
+                    Toast.makeText(context, "USB Device Disconnected", Toast.LENGTH_SHORT).show();
 
-                new PreferenceManager(context).setVersion("none"); // writing version as "none" so that its read againg when PSLab is connected
+                    new PreferenceManager(context).setVersion("none"); // writing version as "none" so that its read againg when PSLab is connected
 
-                if (activityContext != null) {
-                    MainActivity mainActivity = (MainActivity) activityContext;
-                    Fragment currentFragment = mainActivity.getSupportFragmentManager().findFragmentById(R.id.frame);
-                    if (currentFragment instanceof HomeFragment) {
-                        mainActivity.getSupportFragmentManager().beginTransaction().replace(R.id.frame, HomeFragment.newInstance(false, false)).commit();
+                    if (activityContext != null) {
+                        MainActivity mainActivity = (MainActivity) activityContext;
+                        Fragment currentFragment = mainActivity.getSupportFragmentManager().findFragmentById(R.id.frame);
+                        if (currentFragment instanceof HomeFragment) {
+                            mainActivity.getSupportFragmentManager().beginTransaction().replace(R.id.frame, HomeFragment.newInstance(false, false)).commitAllowingStateLoss();
+                        }
                     }
+                } else {
+                    Log.v(TAG, "USB Device is null");
                 }
-            } else {
-                Log.v(TAG, "USB Device is null");
             }
-        }
-    }
+        } catch (IllegalStateException ignored){
 
+        }
+
+    }
 }

